### ISSUE TITLE
[Bug]: Compilation Error with Valid REGEX Pattern Containing `\${..}`

### ISSUE THREAD

**Issue by SachinAkash01** [ballerina-platform/ballerina-lang#41492]
### Description

In Ballerina, when adding REGEX patterns for String constraints, there is a bug that causes a compilation error when using the valid REGEX pattern `\${..}`.  According to the Ballerina specification, ReSyntaxChar: $ can be escaped by using ReQuoteEscape: \

However, applying this valid REGEX pattern results in a compilation error, which is inconsistent with the language specification.

Example:
```ballerina
@constraint:String { pattern: re `^\${2}[a-z]+$`}
```

### Related area

-> Compilation

### Suggested label(s) (optional)

CompilerFE, Regexp

**Comment by LakshanWeerasinghe** [comment]
This pattern is an invalid pattern. The `{` and `}` characters are syntax characters. Which needs to be escaped here. The only issue I see here is the error diagnostics are incorrect. There should be two `missing backslash` diagnostics.

**Comment by LakshanWeerasinghe** [comment]
The correct pattern you want to use would be `^\$\{2\}[a-z]+$`.

**Comment by LakshanWeerasinghe** [comment]
@SasinduDilshara yes, this is quantifier scenario. We need to fix in both compile time and runtime parsers.

**Comment by SachinAkash01** [comment]
After fixing this bug, this improvement [#5048](https://github.com/ballerina-platform/ballerina-standard-library/issues/5048) also must be fixed.
